File: 766-flatten-a-multilevel-doubly-linked-list/flatten-a-multilevel-doubly-linked-list.cpp
/*
// Definition for a Node.
class Node {
public:
    int val;
    Node* prev;
    Node* next;
    Node* child;
};
*/

class Solution {
Node* helper(Node* head){
    if(head==nullptr) return nullptr;
    Node *curr = head;
    while(curr!=nullptr){
        if(curr->child!=NULL){
            Node *temp = curr->next;
            Node *tail = helper(curr->child);
            curr->next = curr->child;
            curr->child->prev=curr;
            // tail->next = temp;
            // temp->prev=tail;
            curr->child=nullptr;
            if (temp != nullptr){
                tail->next = temp;
                temp->prev = tail;
            }
        }
        if(curr->next==NULL) break;

        curr=curr->next;
    }
    return curr;
}

public:
    Node* flatten(Node* head) {
        if(head==NULL) return NULL;
        helper(head);
        return head;
    }
};

File: 766-flatten-a-multilevel-doubly-linked-list/flatten-a-multilevel-doubly-linked-list.java
/*
// Definition for a Node.
class Node {
    public int val;
    public Node prev;
    public Node next;
    public Node child;
};
*/

class Solution {
    Node n;
    public Node flatten(Node head) {
        n=head;
        traverse();

        return head;
    }
    public void traverse(){
        if(n==null){
            return;
        }
        if(n.next==null){
            if(n.child==null) return;
            else{
                n.next = n.child;
                n.child.prev=n;
                n.child=null;
                n = n.next;
                traverse();
            }
             return;
        }
        if(n.child==null){
            n=n.next;
            traverse();
        }
        else{
            Node p=n.next;
            n.next=n.child;
            n.child.prev=n;
            n.child=null;
            n=n.next;
            traverse();
            n.next=p;
            p.prev=n;
            n=n.next;
            traverse();
        }
    }
}